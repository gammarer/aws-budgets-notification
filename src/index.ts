import * as crypto from 'crypto';
import * as cdk from 'aws-cdk-lib';
import * as budgets from 'aws-cdk-lib/aws-budgets';
import * as bot from 'aws-cdk-lib/aws-chatbot';
import * as iam from 'aws-cdk-lib/aws-iam';
//import * as logs from 'aws-cdk-lib/aws-logs';
import * as sns from 'aws-cdk-lib/aws-sns';
import { Construct } from 'constructs';


export interface BudgetsNotificationProps {
  readonly slackWorkspaceId: string;
  readonly slackChannelId: string;
  readonly budgetLimitAmount: number;
  readonly linkedAccounts?: string[];
}

export class BudgetsNotification extends Construct {

  constructor(scope: Construct, id: string, props: BudgetsNotificationProps) {
    super(scope, id);

    // ðŸ‘‡Get current account & region
    const account = cdk.Stack.of(this).account;
    // const region = cdk.Stack.of(this).region;

    // ðŸ‘‡Create random key
    const randomNameKey = crypto.createHash('shake256', { outputLength: 4 })
      .update(`${cdk.Names.uniqueId(scope)}-${cdk.Names.uniqueId(this)}`)
      .digest('hex');

    // ðŸ‘‡Create SNS Topic
    const topic = new sns.Topic(this, 'NotificationTopic', {
      topicName: `budget-notification-${randomNameKey}-topic`,
      displayName: `budget-notification-${randomNameKey}-topic`,
    });
    topic.addToResourcePolicy(new iam.PolicyStatement({
      sid: 'AWSBudgetsSNSPublishingPermissions',
      effect: iam.Effect.ALLOW,
      principals: [
        new iam.ServicePrincipal('budgets.amazonaws.com'),
      ],
      actions: [
        'sns:Publish',
      ],
      resources: [
        topic.topicArn,
      ],
    }));
    topic.addToResourcePolicy(new iam.PolicyStatement({
      sid: 'OwnerSNSActionPermissions',
      effect: iam.Effect.ALLOW,
      principals: [
        new iam.AnyPrincipal(),
      ],
      actions: [
        'sns:GetTopicAttributes',
        'sns:SetTopicAttributes',
        'sns:AddPermission',
        'sns:RemovePermission',
        'sns:DeleteTopic',
        'sns:Subscribe',
        'sns:ListSubscriptionsByTopic',
        'sns:Publish',
        'sns:Receive',
      ],
      resources: [
        topic.topicArn,
      ],
      conditions: {
        StringEquals: {
          'AWS:SourceOwner': account,
        },
      },
    }));

    // ðŸ‘‡Create ChatBot
    new bot.SlackChannelConfiguration(this, 'SlackChannelConfig', {
      slackChannelConfigurationName: `slack-channel-budget-notification-${randomNameKey}-config`,
      slackWorkspaceId: props.slackWorkspaceId,
      slackChannelId: props.slackChannelId,
      //logRetention: logs.RetentionDays.TWO_MONTHS,
      //logRetentionRetryOptions: undefined,
      //logRetentionRole: undefined,
      loggingLevel: bot.LoggingLevel.ERROR,
      notificationTopics: [
        topic,
      ],
      role: new iam.Role(this, 'SlackChannelConfigRole', {
        roleName: `slack-channel-budget-notification-${randomNameKey}-config-role`,
        description: 'slack channel budget notification config role.',
        assumedBy: new iam.ServicePrincipal('chatbot.amazonaws.com'),
        inlinePolicies: {
          'chatbot-policy': new iam.PolicyDocument({
            statements: [
              new iam.PolicyStatement({
                effect: iam.Effect.ALLOW,
                actions: [
                  'cloudwatch:Describe*',
                  'cloudwatch:Get*',
                  'cloudwatch:List*',
                ],
                resources: ['*'],
              }),
            ],
          }),
        },
      }),
    });

    // ðŸ‘‡Common notification with subscribers.
    const notificationsWithSubscribers: Array<budgets.CfnBudget.NotificationWithSubscribersProperty> = [
      {
        notification: {
          notificationType: 'ACTUAL',
          comparisonOperator: 'GREATER_THAN',
          threshold: 60,
          thresholdType: 'PERCENTAGE',
        },
        subscribers: [
          {
            subscriptionType: 'SNS',
            address: topic.topicArn,
          },
        ],
      },
      {
        notification: {
          notificationType: 'ACTUAL',
          comparisonOperator: 'GREATER_THAN',
          threshold: 80,
          thresholdType: 'PERCENTAGE',
        },
        subscribers: [
          {
            subscriptionType: 'SNS',
            address: topic.topicArn,
          },
        ],
      },
      {
        notification: {
          notificationType: 'ACTUAL',
          comparisonOperator: 'GREATER_THAN',
          threshold: 100,
          thresholdType: 'PERCENTAGE',
        },
        subscribers: [
          {
            subscriptionType: 'SNS',
            address: topic.topicArn,
          },
        ],
      },
    ];

    if (props.linkedAccounts && props.linkedAccounts.length >= 1) {
      for (const linkedAccount of props.linkedAccounts) {
        new budgets.CfnBudget(this, `Budget${linkedAccount}`, {
          budget: {
            budgetType: 'COST',
            budgetName: `Monthly usage for ${linkedAccount}.`,
            timeUnit: 'MONTHLY',
            costFilters: {
              LinkedAccount: [
                linkedAccount,
              ],
            },
            budgetLimit: {
              amount: props.budgetLimitAmount,
              unit: 'USD',
            },
          },
          notificationsWithSubscribers,
        });
      }
    } else {
      new budgets.CfnBudget(this, 'Budget', {
        budget: {
          budgetType: 'COST',
          budgetName: 'Monthly usage',
          timeUnit: 'MONTHLY',
          budgetLimit: {
            amount: props.budgetLimitAmount,
            unit: 'USD',
          },
        },
        notificationsWithSubscribers,
      });
    }
  }
}