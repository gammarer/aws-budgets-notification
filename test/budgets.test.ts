import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { BudgetsNotification } from '../src';


describe('BudgetsNotification Testing', () => {

  describe('BudgetsNotification Normal Testing', () => {
    const app = new App();
    const stack = new Stack(app, 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });
    new BudgetsNotification(stack, 'BudgetsNotification', {
      slackWorkspaceId: 'T0XXXX111',
      slackChannelId: 'XXXXXXXX',
      budgetLimitAmount: 50,
    });
    const template = Template.fromStack(stack);
    describe('SNS Topic Testing', () => {
      it('Should have sns topic', () => {
        template.hasResourceProperties('AWS::SNS::Topic', Match.objectEquals({
          DisplayName: Match.stringLikeRegexp('budget-notification-.*-topic'),
          TopicName: Match.stringLikeRegexp('budget-notification-.*-topic'),
        }));
      });
      it('Should have sns topic resource policy of publish from aws budgets', () => {
        template.hasResourceProperties('AWS::SNS::TopicPolicy', {
          PolicyDocument: {
            Version: '2012-10-17',
            Statement: Match.arrayWith([
              Match.objectEquals({
                Sid: 'AWSBudgetsSNSPublishingPermissions',
                Effect: 'Allow',
                Action: 'sns:Publish',
                Principal: {
                  Service: 'budgets.amazonaws.com',
                },
                Resource: {
                  Ref: Match.stringLikeRegexp('BudgetsNotificationNotificationTopic.*'),
                },
              }),
            ]),
          },
        });
      });
      it('Should have sns topic resource policy of sns actions from owner', () => {
        template.hasResourceProperties('AWS::SNS::TopicPolicy', {
          PolicyDocument: {
            Version: '2012-10-17',
            Statement: Match.arrayWith([
              Match.objectEquals({
                Sid: 'OwnerSNSActionPermissions',
                Effect: 'Allow',
                Action: [
                  'sns:GetTopicAttributes',
                  'sns:SetTopicAttributes',
                  'sns:AddPermission',
                  'sns:RemovePermission',
                  'sns:DeleteTopic',
                  'sns:Subscribe',
                  'sns:ListSubscriptionsByTopic',
                  'sns:Publish',
                  'sns:Receive',
                ],
                Principal: {
                  AWS: '*',
                },
                Resource: {
                  Ref: Match.stringLikeRegexp('BudgetsNotificationNotificationTopic.*'),
                },
                Condition: {
                  StringEquals: {
                    'AWS:SourceOwner': '123456789012',
                  },
                },
              }),
            ]),
          },
        });
      });
    });

    describe('ChatBot SlackChannelConfiguration Testing', () => {
      it('Should have SlackChannelConfiguration role', () => {
        template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
          RoleName: Match.stringLikeRegexp('slack-channel-budget-notification-.*-config-role'),
          Description: Match.anyValue(),
          AssumeRolePolicyDocument: Match.objectEquals({
            Version: '2012-10-17',
            Statement: Match.arrayWith([
              Match.objectEquals({
                Effect: 'Allow',
                Principal: {
                  Service: 'chatbot.amazonaws.com',
                },
                Action: 'sts:AssumeRole',
              }),
            ]),
          }),
          Policies: Match.arrayEquals([
            {
              PolicyName: 'chatbot-policy',
              PolicyDocument: Match.objectEquals({
                Version: '2012-10-17',
                Statement: [
                  Match.objectEquals({
                    Effect: 'Allow',
                    Action: [
                      'cloudwatch:Describe*',
                      'cloudwatch:Get*',
                      'cloudwatch:List*',
                    ],
                    Resource: '*',
                  }),
                ],
              }),
            },
          ]),
        }));
      });
      it('Should have SlackChannelConfiguration', () => {
        template.hasResourceProperties('AWS::Chatbot::SlackChannelConfiguration', Match.objectEquals({
          ConfigurationName: Match.stringLikeRegexp('slack-channel-budget-notification-.*-config'),
          IamRoleArn: {
            'Fn::GetAtt': Match.arrayEquals([
              Match.stringLikeRegexp('BudgetsNotificationSlackChannelConfigRole.*'),
              'Arn',
            ]),
          },
          LoggingLevel: 'ERROR',
          SlackWorkspaceId: 'T0XXXX111',
          SlackChannelId: 'XXXXXXXX',
          SnsTopicArns: [
            {
              Ref: Match.stringLikeRegexp('BudgetsNotificationNotificationTopic.*'),
            },
          ],
          //UserRoleRequired: ,
        }));
      });
    });
    describe('Budget Testing', () => {
      it('Should have Budget', () => {
        template.hasResourceProperties('AWS::Budgets::Budget', Match.objectEquals({
          Budget: {
            BudgetName: 'Monthly usage',
            BudgetType: 'COST',
            TimeUnit: 'MONTHLY',
            BudgetLimit: {
              Amount: 50,
              Unit: 'USD',
            },
          },
          NotificationsWithSubscribers: Match.arrayWith([
            Match.objectEquals({
              Notification: Match.objectEquals({
                ComparisonOperator: 'GREATER_THAN',
                NotificationType: 'ACTUAL',
                Threshold: 60,
                ThresholdType: 'PERCENTAGE',
              }),
              Subscribers: Match.arrayWith([
                Match.objectEquals({
                  Address: {
                    Ref: Match.stringLikeRegexp('BudgetsNotificationNotificationTopic.*'),
                  },
                  SubscriptionType: 'SNS',
                }),
              ]),
            }),
            Match.objectEquals({
              Notification: Match.objectEquals({
                ComparisonOperator: 'GREATER_THAN',
                NotificationType: 'ACTUAL',
                Threshold: 80,
                ThresholdType: 'PERCENTAGE',
              }),
              Subscribers: Match.arrayWith([
                Match.objectEquals({
                  Address: {
                    Ref: Match.stringLikeRegexp('BudgetsNotificationNotificationTopic.*'),
                  },
                  SubscriptionType: 'SNS',
                }),
              ]),
            }),
            Match.objectEquals({
              Notification: Match.objectEquals({
                ComparisonOperator: 'GREATER_THAN',
                NotificationType: 'ACTUAL',
                Threshold: 100,
                ThresholdType: 'PERCENTAGE',
              }),
              Subscribers: Match.arrayWith([
                Match.objectEquals({
                  Address: {
                    Ref: Match.stringLikeRegexp('BudgetsNotificationNotificationTopic.*'),
                  },
                  SubscriptionType: 'SNS',
                }),
              ]),
            }),
          ]),
        }));
      });
    });
    it('Should match snapshot', () => {
      expect(template.toJSON()).toMatchSnapshot('budgets-normal');
    });
  });

  describe('BudgetsNotification LinkedAccounts Testing', () => {
    const app = new App();
    const stack = new Stack(app, 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });
    new BudgetsNotification(stack, 'BudgetsNotification', {
      slackWorkspaceId: 'T0XXXX111',
      slackChannelId: 'XXXXXXXX',
      budgetLimitAmount: 50,
      linkedAccounts: [
        '111111111111',
        '222222222222',
      ],
    });
    const template = Template.fromStack(stack);

    it('Should match budgets resource count', () => {
      template.resourceCountIs('AWS::Budgets::Budget', 2);
    });

    it('Should match snapshot', () => {
      expect(template.toJSON()).toMatchSnapshot('budgets-linked-accounts');
    });
  });
});
