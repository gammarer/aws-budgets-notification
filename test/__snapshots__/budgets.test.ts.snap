// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetsNotification Testing BudgetsNotification LinkedAccounts Testing Should match snapshot: budgets-linked-accounts 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "BudgetsNotificationBudget11111111111188E95F35": Object {
      "Properties": Object {
        "Budget": Object {
          "BudgetLimit": Object {
            "Amount": 50,
            "Unit": "USD",
          },
          "BudgetName": "Monthly usage for 111111111111.",
          "BudgetType": "COST",
          "CostFilters": Object {
            "LinkedAccount": Array [
              "111111111111",
            ],
          },
          "TimeUnit": "MONTHLY",
        },
        "NotificationsWithSubscribers": Array [
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 60,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
                },
                "SubscriptionType": "SNS",
              },
            ],
          },
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 80,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
                },
                "SubscriptionType": "SNS",
              },
            ],
          },
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 100,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
                },
                "SubscriptionType": "SNS",
              },
            ],
          },
        ],
      },
      "Type": "AWS::Budgets::Budget",
    },
    "BudgetsNotificationBudget22222222222281269B64": Object {
      "Properties": Object {
        "Budget": Object {
          "BudgetLimit": Object {
            "Amount": 50,
            "Unit": "USD",
          },
          "BudgetName": "Monthly usage for 222222222222.",
          "BudgetType": "COST",
          "CostFilters": Object {
            "LinkedAccount": Array [
              "222222222222",
            ],
          },
          "TimeUnit": "MONTHLY",
        },
        "NotificationsWithSubscribers": Array [
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 60,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
                },
                "SubscriptionType": "SNS",
              },
            ],
          },
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 80,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
                },
                "SubscriptionType": "SNS",
              },
            ],
          },
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 100,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
                },
                "SubscriptionType": "SNS",
              },
            ],
          },
        ],
      },
      "Type": "AWS::Budgets::Budget",
    },
    "BudgetsNotificationNotificationTopicDFB74E64": Object {
      "Properties": Object {
        "DisplayName": "budget-notification-586e7973-topic",
        "TopicName": "budget-notification-586e7973-topic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "BudgetsNotificationNotificationTopicPolicyD5C1A9D0": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "budgets.amazonaws.com",
              },
              "Resource": Object {
                "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
              },
              "Sid": "AWSBudgetsSNSPublishingPermissions",
            },
            Object {
              "Action": Array [
                "sns:GetTopicAttributes",
                "sns:SetTopicAttributes",
                "sns:AddPermission",
                "sns:RemovePermission",
                "sns:DeleteTopic",
                "sns:Subscribe",
                "sns:ListSubscriptionsByTopic",
                "sns:Publish",
                "sns:Receive",
              ],
              "Condition": Object {
                "StringEquals": Object {
                  "AWS:SourceOwner": "123456789012",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Object {
                "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
              },
              "Sid": "OwnerSNSActionPermissions",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": Array [
          Object {
            "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
    "BudgetsNotificationSlackChannelConfigBE947213": Object {
      "Properties": Object {
        "ConfigurationName": "slack-channel-budget-notification-586e7973-config",
        "IamRoleArn": Object {
          "Fn::GetAtt": Array [
            "BudgetsNotificationSlackChannelConfigRole34A2F398",
            "Arn",
          ],
        },
        "LoggingLevel": "ERROR",
        "SlackChannelId": "XXXXXXXX",
        "SlackWorkspaceId": "T0XXXX111",
        "SnsTopicArns": Array [
          Object {
            "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
          },
        ],
      },
      "Type": "AWS::Chatbot::SlackChannelConfiguration",
    },
    "BudgetsNotificationSlackChannelConfigRole34A2F398": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "chatbot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "slack channel budget notification config role.",
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "cloudwatch:Describe*",
                    "cloudwatch:Get*",
                    "cloudwatch:List*",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "chatbot-policy",
          },
        ],
        "RoleName": "slack-channel-budget-notification-586e7973-config-role",
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`BudgetsNotification Testing BudgetsNotification Normal Testing Should match snapshot: budgets-normal 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "BudgetsNotificationBudgetB79C7685": Object {
      "Properties": Object {
        "Budget": Object {
          "BudgetLimit": Object {
            "Amount": 50,
            "Unit": "USD",
          },
          "BudgetName": "Monthly usage",
          "BudgetType": "COST",
          "TimeUnit": "MONTHLY",
        },
        "NotificationsWithSubscribers": Array [
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 60,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
                },
                "SubscriptionType": "SNS",
              },
            ],
          },
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 80,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
                },
                "SubscriptionType": "SNS",
              },
            ],
          },
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 100,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
                },
                "SubscriptionType": "SNS",
              },
            ],
          },
        ],
      },
      "Type": "AWS::Budgets::Budget",
    },
    "BudgetsNotificationNotificationTopicDFB74E64": Object {
      "Properties": Object {
        "DisplayName": "budget-notification-586e7973-topic",
        "TopicName": "budget-notification-586e7973-topic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "BudgetsNotificationNotificationTopicPolicyD5C1A9D0": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "budgets.amazonaws.com",
              },
              "Resource": Object {
                "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
              },
              "Sid": "AWSBudgetsSNSPublishingPermissions",
            },
            Object {
              "Action": Array [
                "sns:GetTopicAttributes",
                "sns:SetTopicAttributes",
                "sns:AddPermission",
                "sns:RemovePermission",
                "sns:DeleteTopic",
                "sns:Subscribe",
                "sns:ListSubscriptionsByTopic",
                "sns:Publish",
                "sns:Receive",
              ],
              "Condition": Object {
                "StringEquals": Object {
                  "AWS:SourceOwner": "123456789012",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Object {
                "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
              },
              "Sid": "OwnerSNSActionPermissions",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": Array [
          Object {
            "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
    "BudgetsNotificationSlackChannelConfigBE947213": Object {
      "Properties": Object {
        "ConfigurationName": "slack-channel-budget-notification-586e7973-config",
        "IamRoleArn": Object {
          "Fn::GetAtt": Array [
            "BudgetsNotificationSlackChannelConfigRole34A2F398",
            "Arn",
          ],
        },
        "LoggingLevel": "ERROR",
        "SlackChannelId": "XXXXXXXX",
        "SlackWorkspaceId": "T0XXXX111",
        "SnsTopicArns": Array [
          Object {
            "Ref": "BudgetsNotificationNotificationTopicDFB74E64",
          },
        ],
      },
      "Type": "AWS::Chatbot::SlackChannelConfiguration",
    },
    "BudgetsNotificationSlackChannelConfigRole34A2F398": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "chatbot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "slack channel budget notification config role.",
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "cloudwatch:Describe*",
                    "cloudwatch:Get*",
                    "cloudwatch:List*",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "chatbot-policy",
          },
        ],
        "RoleName": "slack-channel-budget-notification-586e7973-config-role",
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
